resource_retriever/src/plugins/curl_retriever.cpp

```diff
@@ -0,0 +1,157 @@
+// Copyright 2009, Willow Garage, Inc. All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are met:
+//
+//    * Redistributions of source code must retain the above copyright
+//      notice, this list of conditions and the following disclaimer.
+//
+//    * Redistributions in binary form must reproduce the above copyright
+//      notice, this list of conditions and the following disclaimer in the
+//      documentation and/or other materials provided with the distribution.
+//
+//    * Neither the name of the  Willow Garage, Inc. nor the names of its
+//      contributors may be used to endorse or promote products derived from
+//      this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+#include "resource_retriever/plugins/curl_retriever.hpp"
+
+#include <curl/curl.h>
+
+#include <array>
+#include <cstring>
+#include <iostream>
+#include <memory>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "resource_retriever/exception.hpp"
+
+namespace
+{
+
+class CURLStaticInit
+{
+public:
+  CURLStaticInit()
+  {
+    CURLcode ret = curl_global_init(CURL_GLOBAL_ALL);
+    if (ret != 0) {
+      std::cerr << "Error initializing libcurl! retcode = " << ret;
```

Comment:
Should we throw an exception here?  That way we wouldn't need the `initialized_` variable at all.

Reply:
I think the issue is that this is used a static global, so it's not possible to catch the exception.

However, I agree this isn't great.

I considered at least saying it should go through some injectable logging (so we could pipe it to ROS logging for example), but that's hard too since this is a global.

I think maybe it shouldn't print anything and instead it should just set the flag and the first place to try and use it should throw an exception since it's not initialized.

But I'll have to think about it more.

Reply:
I reworked this to throw an exception the first time we attempt to use curl rather than print an error here.